src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -397,6 +400,43 @@ where
                             }
                         }
                     }
+
+                    // Before processing barrier, if did not snapshot read,
+                    // do a snapshot read first.
+                    // This is so we don't lose the tombstone iteration progress.
+                    if !has_snapshot_read {
+                        let (_, snapshot) = backfill_stream.into_inner();
+                        #[for_await]
+                        for msg in snapshot {
+                            let Either::Right(msg) = msg else {
+                                bail!("BUG: snapshot_read contains upstream messages");
+                            };
+                            match msg? {
+                                None => {
+                                    // End of the snapshot read stream.
+                                    // We let the barrier handling logic take care of upstream updates.
+                                    // But we still want to exit backfill loop, so we mark snapshot read complete.
+                                    snapshot_read_complete = true;
+                                    break;
+                                }
+                                Some(chunk) => {
```

Comment:
BTW, since in this PR, we wait a least a chunk it seems we can remove the chunk builder for snapshot read as well though it will waste some resources when the chunk builder is not full.

Reply:
OR as an improvement, if we keep the chunk builder, We can wait one row instead of a chunk to avoid stalling the barrier too long.

Reply:
When we break out of this branch, we will enter the barrier handling branch.
There we will consume the remaining rows.

I think it is hard to wait 1 row, since the stream itself uses chunks as the iteration unit.